core/trino-main/src/main/java/io/trino/sql/ir/IrUtils.java

```diff
@@ -43,6 +45,11 @@ public final class IrUtils
 {
     private IrUtils() {}
 
+    static void validateType(Type expected, Expression expression)
+    {
+        checkArgument(expected.equals(expression.type()), "Expected '%s' type but found '%s' for expression: %s", expected, expression.type(), expression);
```

Comment:
since you've removed rnn from the record constructors maybe we should add it here?

```
requireNonNull(expected, "expected is null");
requireNonNull(expression, "expression is null");
```

Reply:
Not needed. The check will blow up if they are null, and with a better error message